File: NeetCode/Arrays_and_Hashing/Range_Sum_Query_2D_Immutable/range__sum__query_2_d__immutable.py
from typing import Optional, List
import heapq
from collections import defaultdict


class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        self.height = len(matrix)
        self.width = len(matrix[0])
        self.sumMatrix = [
            [0 for _ in range(self.width + 1)] for _ in range(self.height + 1)
        ]
        for y in range(self.height):
            for x in range(self.width):
                self.sumMatrix[y + 1][x + 1] = (
                    matrix[y][x]
                    + self.sumMatrix[y + 1][x]
                    + self.sumMatrix[y][x + 1]
                    - self.sumMatrix[y][x]
                )

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        return (
            self.sumMatrix[row2 + 1][col2 + 1]
            - self.sumMatrix[row1][col2 + 1]
            - self.sumMatrix[row2 + 1][col1]
            + self.sumMatrix[row1][col1]
        )


if __name__ == "__main__":
    matrix = [
        [3, 0, 1, 4, 2],
        [5, 6, 3, 2, 1],
        [1, 2, 0, 1, 5],
        [4, 1, 0, 1, 7],
        [1, 0, 3, 0, 5],
    ]
    nm = NumMatrix(matrix)
    print(nm.sumRegion(2, 1, 4, 3))
    print("Running Solution...")

File: NeetCode/Arrays_and_Hashing/Range_Sum_Query_2D_Immutable/Range_Sum_Query_2D_Immutable.cpp
// Range_Sum_Query_2D_Immutable.cpp
//
// Compile locally with:
// clang++ Range_Sum_Query_2D_Immutable.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class NumMatrix
{
private:
    vector<vector<int>> sumMatrix;
    int height;
    int width;

public:
    NumMatrix(vector<vector<int>> &matrix)
    {
        height = matrix.size();
        width = matrix[0].size();
        sumMatrix = vector<vector<int>>(height + 1, vector<int>(width + 1, 0));
        for (int y = 0; y < height; y++)
        {
            for (int x = 0; x < width; x++)
            {
                sumMatrix[y + 1][x + 1] = matrix[y][x] + sumMatrix[y][x + 1] + sumMatrix[y + 1][x] - sumMatrix[y][x];
            }
        }
    }

    int sumRegion(int row1, int col1, int row2, int col2)
    {
        return sumMatrix[row2 + 1][col2 + 1] - sumMatrix[row2 + 1][col1] - sumMatrix[row1][col2 + 1] + sumMatrix[row1][col1];
    }
};

int main()
{
    cout << "Running Range_Sum_Query_2D_Immutable.cpp..." << endl;

    // TODO:
    // Add local test calls here

    return 0;
}
